File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean dfs(TreeNode cur, int target, int sum){
        if (cur == null) return false;
        if (cur.left == null && cur.right == null){
            if (target == sum + cur.val) return true;
            else return false;
        };
        return dfs(cur.left,target,sum + cur.val) || dfs(cur.right, target, sum + cur.val);
    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
}

File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool dfs(TreeNode* root, int sum, int target) {
        if (root == nullptr) return false;
        if (root->left == nullptr && root->right == nullptr){
            sum += root->val;
            if (sum == target) return true;
            return false;
        }
        return dfs(root->left, sum + root->val, target) |
               dfs(root->right, sum + root->val, target);
    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (root == nullptr) return false;
        return dfs(root, 0, targetSum);
    }
};